tools/cbmc/proofs/HTTP/global_state_HTTP.c

```diff
@@ -123,103 +183,193 @@ int is_valid_NetworkInterface(IotNetworkInterface_t *netif) {
-    __CPROVER_assume(pResponseHandle->readHeaderValueLength >= 0 &&
-    pResponseHandle->readHeaderValueLength <= (pResponseHandle->pHeadersEnd - pResponseHandle->pHeaders));
-    pResponseHandle->pReadHeaderValue = malloc(pResponseHandle->readHeaderValueLength);
+
+    pResponseHandle->pBody = safeMalloc(len);
+    pResponseHandle->pHttpsConnection = allocate_IotConnectionHandle();
+    pResponseHandle->pReadHeaderField =
+      safeMalloc(pResponseHandle->readHeaderFieldLength);
+    pResponseHandle->pReadHeaderValue =
+      safeMalloc(pResponseHandle->readHeaderValueLength);
   }
   return pResponseHandle;
 }
 
+int is_valid_IotResponseHandle(IotHttpsResponseHandle_t pResponseHandle) {
+  int required =
+    __CPROVER_same_object(pResponseHandle->pHeaders,
+			  pResponseHandle->pHeadersCur) &&
+    // ???: Does this overconstrain End?
+    __CPROVER_same_object(pResponseHandle->pHeaders,
+			  pResponseHandle->pHeadersEnd);
+  if (!required) return 0;
+
+  int valid_headers =
+    pResponseHandle->pHeaders == ((_resHandle_t*)pResponseHandle)->data;
+  int valid_order =
+    pResponseHandle->pHeaders <= pResponseHandle->pHeadersCur &&
+    pResponseHandle->pHeadersCur <=  pResponseHandle->pHeadersEnd;
+  int valid_parserdata =
+    pResponseHandle->httpParserInfo.readHeaderParser.data == pResponseHandle;
+  return
+    valid_headers &&
+    valid_order &&
+    valid_parserdata;
+}
+
+
+/****************************************************************
+ * IotHttpsRequestHandle constructor
+ ****************************************************************/
+
 /* Creates a Request Handle and assigns memory accordingly. */
-IotHttpsRequestHandle_t newIotRequestHandle() {
+IotHttpsRequestHandle_t allocate_IotRequestHandle() {
   IotHttpsRequestHandle_t pRequestHandle = safeMalloc(sizeof(_reqHandle_t));
   if (pRequestHandle) {
     uint32_t len;
-    pRequestHandle->pHttpsResponse = newIotResponseHandle();
-    pRequestHandle->pHttpsConnection = newIotConnectionHandle();
-    pRequestHandle->pBody = malloc(len);
-    pRequestHandle->pHeaders = ((_reqHandle_t*)pRequestHandle)->data;
-    pRequestHandle->pHeadersCur = pRequestHandle->pHeaders;
-    pRequestHandle->pHeadersEnd = pRequestHandle->pHeaders + sizeof(((_reqHandle_t*)pRequestHandle)->data);
-    pRequestHandle->pConnInfo = newConnectionInfo();
+    pRequestHandle->pHttpsResponse = allocate_IotResponseHandle();
+    pRequestHandle->pHttpsConnection = allocate_IotConnectionHandle();
+    pRequestHandle->pBody = safeMalloc(len);
+    pRequestHandle->pConnInfo = allocate_IotConnectionInfo();
   }
   return pRequestHandle;
 }
+
+int is_valid_IotRequestHandle(IotHttpsRequestHandle_t pRequestHandle) {
+  int required =
+    __CPROVER_same_object(pRequestHandle->pHeaders,
+			  pRequestHandle->pHeadersCur) &&
+    // ???: Does this overconstrain End?
+    __CPROVER_same_object(pRequestHandle->pHeaders,
+			  pRequestHandle->pHeadersEnd);
```

Comment:
pRequestHandle->pHeadersEnd needs to be the end of the buffer ((_reqHandle_t*)pRequestHandle)->data.
So it must be equal to (((_reqHandle_t*)pRequestHandle)->data + USER_DATA_SIZE)